### PR TITLE
DL3-35 secure middleware courses endpoint

### PR DESCRIPTION
**Pull request by mgwozdz-unicon**
## Description
Secured the middleware's backend endpoint for retrieving the Harmony course catalog using the LTI id_token.

### Motivation and Context
Only users who have a valid LTI registration who are accessing the Deep Linking menu through a valid LTI flow should be permitted to access the course catalog via the middleware.

### Fixes
[DL3-35](https://lumenlearning.atlassian.net/browse/DL3-35)

## How Has This Been Tested?
0. Ensure as a prerequisite that the tool was registered and deployed in D2L and that a Deep Linking Quicklink Link has been added to the Deployment and that there is an org unit appropriately given permission to access the tool.
1. Go to the course (org unit) that has permission to access the tool in D2L.
2. Click on Content in the top navigation.
3. Click "Add a module...", write a name for it, and hit enter to create the module.
4. Click the dropdown for Existing Activities
5. Click on the Deep Linking Quicklink (e.g. I named mine "Lumen Content Selector").
6. Ensure that the course catalog displays with the expected data being sent from the Harmony /course_catalog endpoint and that the pagination is in place if there are more than 10 courses.
Note: If all of the courses in the catalog are using the brain image and were released on 1/18/2022, then this is an indication that the call to Harmony failed and test data has been returned.

## Screenshots
N/A
---
## Database changes
N/A

## ⚠️ Deployment instructions ⚠️
N/A

## New ENV variables
N/A

---

## Checklist
- [x] I have reviewed the AC for this feature and my changes meet all requirements
- [x] (For UI changes) I have considered the accessibility of this feature (see https://www.a11yproject.com/checklist/ for a high-level checklist)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have requested a review from at least one other dev

### CODE REVIEW

**Review by leannethng — APPROVED** [review]
Thank you Mary!

File: src/main/frontend/src/app/appSlice.js [hunk 1]
```diff
@@ -20,13 +32,25 @@ export const appSlice = createSlice({
       state.metadata = action.payload;
     },
     changeSearchInput: (state, action) => {
+      debugger;
```
Comment (smith750) [inline src/main/frontend/src/app/appSlice.js:35]: I'm assuming this line will be removed before the PR is out of draft?

File: src/main/java/net/unicon/lti/controller/harmony/HarmonyRestController.java [hunk 2]
```diff
@@ -45,26 +36,24 @@ public class HarmonyRestController {
...
+    public ResponseEntity<HarmonyPageResponse> listHarmonyCourses(@RequestHeader(value="lti-id-token") String ltiIdToken) {
+        //To keep this endpoint secured, we will validate the id_token
+        try {
+            // validates JWT signature, ensures existing platformDeployment, validates 1.3 format of JWT, validates nonce
+            lti3Request = new LTI3Request(ltiDataService, true, null, ltiIdToken);
+            log.debug("The id_token is valid, fetching courses....");
+            // We convert the JSON response to a Java object, and send the JSON value again to the frontend.
+            return ResponseEntity.ok(harmonyService.fetchHarmonyCourses());
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (smith750) [inline src/main/java/net/unicon/lti/controller/harmony/HarmonyRestController.java:53]: Should this go through the logger?

**Review by smith750 — APPROVED** [review]
I made some nit-picky comments but this looks fantastic.  Thank you!

**Review by HotchkissLuLe — APPROVED** [review]
This looks great, thanks Mary!

File: src/main/frontend/src/app/appSlice.js [hunk 3]
```diff
@@ -50,7 +74,10 @@ export const selectLoading = (state) => state.loading;
 export const fetchCourses = () => (dispatch) => {
   // We must display an spinner when loading courses from the backend
   dispatch(setLoading(true));
-  fetch('/harmony/courses')
+  fetch('/harmony/courses', {
+    method: 'GET',
+    headers: {'lti-id-token': initialState.id_token}
```
Comment (mpellicer) [inline src/main/frontend/src/app/appSlice.js:79]: We can use the state, state.id_token

File: src/main/frontend/src/app/appSlice.js [hunk 4]
```diff
@@ -1,12 +1,24 @@
...
+  selectedCourse: null,
+  courseArray: STATIC_COURSE_ARRAY,
+  id_token: ltiLaunchData.id_token
+};
 
 // This defines the store slice
 export const appSlice = createSlice({
   // Name of the slice
   name: 'appSlice',
+  // Initial state
```
Comment (mpellicer) [inline src/main/frontend/src/app/appSlice.js:20]: The initial state is defined in index.js and it's already inserting the id_token

File: src/main/frontend/src/app/appSlice.js [hunk 5]
```diff
@@ -1,12 +1,24 @@
 import { createSlice } from '@reduxjs/toolkit';
+import { STATIC_COURSE_ARRAY } from './staticData.js';
 
-// Static data
-import { STATIC_COURSE_ARRAY } from './staticData';
+// Parse the LtiLaunchData JSON object from the root attribute.
+const ltiLaunchData = JSON.parse(document.getElementById('root').getAttribute('lti-launch-data'));
```
Comment (mpellicer) [inline src/main/frontend/src/app/appSlice.js:5]: This is done in the index.js, we can use the token.

File: src/main/frontend/src/app/appSlice.js [hunk 6]
```diff
@@ -1,12 +1,24 @@
 import { createSlice } from '@reduxjs/toolkit';
+import { STATIC_COURSE_ARRAY } from './staticData.js';
```
Comment (mpellicer) [inline src/main/frontend/src/app/appSlice.js:2]: This is not needed, in DL3-34 it simulates pagination and is updated.

**Comment by mgwozdz-unicon** [comment]
I believe I've addressed all of the comments and set this as ready for review. Let me know if there are any additional changes I should make. Thank you!

**Comment by smith750** [comment]
Thank you for the updates Mary!

**Review by mpellicer — APPROVED** [review]

**Pull request merged**